lib/rucio/core/transfer.py

```diff
@@ -1063,6 +1064,88 @@ def apply(self, ctx: RequestRankingContext, source: RequestSource) -> "Optional[
         return path[0].src.distance
 
 
+class TransferWaitTime(SourceRankingStrategy):
+    """
+    A source ranking strategy that ranks source nodes based on the expected wait time in queue with exploration embedded
+    through randomization. Note that smaller expected wait times are better.
+    """
+    # There is ~10 s of overhead for any file transfer
+    FILE_OVERHEAD_MS = 10000
+    # Assume 10 Mbps transfer speed
+    PER_BYTE_MS = 1000 / ((10 ** 7) / 8)
+    # Probability we choose an above-average transfer time source node
+    ABOVE_AVG_EXPLORATION = 0.1
+
+    class _RankingContext(RequestRankingContext):
+        def __init__(
+                self,
+                strategy: "TransferWaitTime",
+                rws: "RequestWithSources",
+                costs: "Mapping[RseData, int]"
+        ) -> None:
+            super().__init__(strategy, rws)
+            self.costs = costs
+
+    def for_request(
+            self,
+            rws: RequestWithSources,
+            sources: "Iterable[RequestSource]",
+            *,
+            logger: "LoggerFunction" = logging.log,
+            session: "Session"
+    ) -> "RequestRankingContext":
+        wait_by_rse = {}
+
+        # For each source, estimate the average queue waiting time
+        for src in sources:
+            src_files_queued = 0
+            src_bytes_queued = 0
+            metrics = request_core.get_request_metrics(src_rse_id=src.rse.id, session=session)
+
+            for payload in metrics.values():
+                src_files_queued += payload.get('files', {}).get('queued-total', 0)
+                src_bytes_queued += payload.get('bytes', {}).get('queued-total', 0)
+
+            wait_by_rse[src.rse] = src_files_queued * self.FILE_OVERHEAD_MS + src_bytes_queued * self.PER_BYTE_MS
+
+        costs = defaultdict(lambda: 0)
```

Comment:
defaultdict(int) or float seems more appropriate. Even if i'm not sure it will actually be faster on python interpreter side.